File: 2015/01.js
const floor = s => s === '(' ? 1 : -1
const sum2Basement = (acc, s, i) => {
  const prev = i > 0 ? acc[i - 1] : 0
  if (prev === -1) return false
  acc.push(prev + floor(s))
  return acc
}

const part1 = data => data.sum(floor)
const part2 = data => data.transform(sum2Basement, []).length

export default () => ({ 
  separator: '',
  part1, part2 
})

File: 2015/01.hs
module Main where

import Data.List (foldl')

part1 :: IO ()
part1 = readFile "01.input" >>= print . last . calcFloor

part2 :: IO ()
part2 = readFile "01.input" >>= print . reachesBasementPosition

reachesBasementPosition :: String -> Int
reachesBasementPosition = length . takeWhile (/= -1) . calcFloor

calcFloor :: String -> [Int]
calcFloor s = scanl step 0 s
  where step c '(' = c + 1
        step c ')' = c - 1